File: src/completed/python/euler0014.py
###
#
# Created by Holden on 12/17/2015
#
# SOLVED on 1/6/2016
#
# Problem 14 - Longest Collatz Sequence
#
###

def sequence_length(start):
    seq_length = 1
    
    while start > 1:
        if start % 2 != 0:
            start = (start * 3) + 1
            seq_length += 1
        else:
            start = start // 2
            seq_length += 1
    
    return seq_length

def longest_collatz_sequence(upper):
    long_lenth = 0
    init = 0
    
    for i in range (1, upper):
        seq_length = sequence_length(i)
        
        if seq_length > long_lenth:
            long_lenth = seq_length
            init = i
    
    return init

File: src/completed/java/euler0014.java
/**
 * Created by Holden on 12/02/2016. - COMPLETED
 *
 * Problem 14 - Longest Collatz Sequence
 *
 */

public class Prob0014 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        int answer = longestCollatzSequence(1000000);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static int longestCollatzSequence(int upperBound) {
        int longest = 0;
        int number = 0;
        int length;

        for (int i = 1; i < upperBound; i++) {
            length = sequenceLength(i);
            if (length > longest) {
                longest = length;
                number = i;
            }
        }

        return number;
    }

    private static int sequenceLength(long start) {
        int length = 1;

        while (start > 1) {
            if (start % 2 != 0) {
                start = (start * 3) + 1;
                length++;
            } else {
                start /= 2;
                length++;
            }
        }

        return length;
    }
}
